pennylane/workflow/execution.py

```diff
@@ -160,6 +160,14 @@ def cost_fn(params, x):
     if not isinstance(device, qml.devices.Device):
         device = qml.devices.LegacyDeviceFacade(device)
 
+    # If the transform program contains set_shots, we need to adjust the tape
+    if transform_program is not None and any(
+        getattr(getattr(t, "_transform"), "__name__", "") == "set_shots"
+        for t in getattr(transform_program, "_transform_program", [])
+    ):
+        tapes, _ = transform_program(tapes)
+        transform_program = None
+

```

Comment:
@albi3ro This is an experiemntal draft idea about how to make them compatible. Looks like maintaining the `tapes` at the beginning is sufficient for the usecase in test to run. If this idea sounds proper I can go further to refine it here, e.g. avoid direct calling private attr and go through all `set_shots` in the user transform programs.